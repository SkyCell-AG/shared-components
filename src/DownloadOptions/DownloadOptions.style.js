import {
    makeStyles,
} from '@mui/styles'

const useStyles = makeStyles(() => {
    return {
        containerBtn: {
            display: 'flex',
            justifyContent: 'center',
            flexDirection: 'column',
            alignItems: 'center',
            '@media print': {
                display: 'none',
            },
        },
        btnControl: {
            paddingTop: 20,
        },
    }
})

export default useStyles
