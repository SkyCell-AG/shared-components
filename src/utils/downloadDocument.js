const downloadDocument = (response, name) => {
    const contentType = response.headers['content-type']
    const file = new Blob([response.data], {
        type: contentType,
    })

    const url = window.URL.createObjectURL(file)
    const link = document.createElement('a')

    link.href = url
    const fileName = `${name}.${contentType.split('/')[1]}`

    link.setAttribute('download', fileName)
    document.body.appendChild(link)
    link.click()
    link.remove()
}

export default downloadDocument
