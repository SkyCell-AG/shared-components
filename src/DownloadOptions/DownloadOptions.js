import React, {
    useCallback,
} from 'react'
import PropTypes from 'prop-types'
import noop from 'lodash/noop'

import {
    dateToStr,
    dateTimeMask,
} from '../utils/DateUtils'
import downloadDocument from '../utils/downloadDocument'

import Button from '../Button'
import Card from '../Card'
import Radio from '../Radio'
import TemperatureRange from '../TemperatureRange'

import useStyles from './DownloadOptions.style'

const propTypes = {
    showTemperatureRangeOption: PropTypes.bool,
    showPdfButton: PropTypes.bool,
    showCsvButton: PropTypes.bool,
    showTemperatureRangeAllOptions: PropTypes.bool,
    sensorData: PropTypes.arrayOf(PropTypes.any),
    showTempRange: PropTypes.bool.isRequired,
    onCheckShowTempRange: PropTypes.func.isRequired,
    serialNumber: PropTypes.string,
    printChart: PropTypes.func,
    selectedTemperatureRange: PropTypes.string,
    radioOptions: PropTypes.arrayOf(PropTypes.shape({
        label: PropTypes.string,
        value: PropTypes.string,
    })),
    onChangeSelectedTemperatureRange: PropTypes.func,
}

const defaultProps = {
    showTemperatureRangeOption: false,
    showPdfButton: false,
    showCsvButton: false,
    showTemperatureRangeAllOptions: false,
    sensorData: undefined,
    serialNumber: '',
    selectedTemperatureRange: '',
    radioOptions: [],
    printChart: noop,
    onChangeSelectedTemperatureRange: noop,
}

const DownloadOptions = (props) => {
    const classes = useStyles()

    const {
        showTemperatureRangeOption,
        showPdfButton,
        showCsvButton,
        showTemperatureRangeAllOptions,
        serialNumber,
        showTempRange,
        onCheckShowTempRange,
        sensorData,
        printChart,
        selectedTemperatureRange,
        radioOptions,
        onChangeSelectedTemperatureRange,
    } = props

    const exportCsv = useCallback(() => {
        const csvContent = `${sensorData.map((e) => {
            const element = [
                serialNumber,
                dateToStr(e[0], dateTimeMask),
                e[1],
                e[4],
            ]

            return element.join(',')
        }).join('\n')}`

        downloadDocument(
            {
                headers: {
                    'content-type': 'text/csv',
                },
                data: `SERIAL_NUMBER,TIMESTAMP,AMBIENT_TEMPERATURE,INTERNAL_TEMPERATURE\n${csvContent}`,
            },
            `sensor_data_${serialNumber}`,
        )
    }, [
        sensorData,
        serialNumber,
    ])

    return (
        <Card title="Options">
            {showTemperatureRangeOption && (
                <TemperatureRange
                    showTempRange={showTempRange}
                    onCheckShowTempRange={onCheckShowTempRange}
                    title="Temperature Range"
                />
            ) }
            {showTemperatureRangeAllOptions && (
                <Radio
                    className={classes.selectedTemperatureRange}
                    name="selectedTemperatureRange"
                    compact
                    value={selectedTemperatureRange}
                    options={radioOptions}
                    onChange={onChangeSelectedTemperatureRange}
                />
            )}
            <div className={classes.containerBtn}>
                {showPdfButton && (
                    <Button
                        className={classes.btnControl}
                        label="pdf"
                        isUpperCase
                        onClick={printChart}
                    />
                )}
                {showCsvButton && (
                    <Button
                        className={classes.btnControl}
                        label="csv"
                        isUpperCase
                        onClick={exportCsv}
                    />
                )}
            </div>
        </Card>
    )
}

DownloadOptions.propTypes = propTypes
DownloadOptions.defaultProps = defaultProps

export default DownloadOptions
