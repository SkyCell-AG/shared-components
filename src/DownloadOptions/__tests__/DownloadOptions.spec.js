import React from 'react'
import {
    shallow,
} from 'enzyme'

import downloadDocument from '../../utils/downloadDocument'

import DownloadOptions from '../DownloadOptions'

jest.mock('../../utils/downloadDocument')

describe('shared-components/DownloadOptions', () => {
    it('should match snapshot', () => {
        const wrapper = shallow(
            <DownloadOptions
                showTempRange
                onCheckShowTempRange={jest.fn()}
            />,
        )

        expect(wrapper).toMatchSnapshot()
    })

    it('should match snapshot with all buttons', () => {
        const wrapper = shallow(
            <DownloadOptions
                showPdfButton
                showCsvButton
                showTemperatureRangeOption
                showTemperatureRangeAllOptions
                showTempRange
                onCheckShowTempRange={jest.fn()}
            />,
        )

        expect(wrapper).toMatchSnapshot()
    })

    it('Document Print is loaded', (done) => {
        const wrapper = shallow(
            <DownloadOptions
                showCsvButton
                showTempRange
                serialNumber="031-10279"
                sensorData={[[
                    new Date('August 16, 2018 06:50:00'),
                    24.7,
                    null,
                    null,
                    21.5,
                    null,
                    null,
                ]]}
                onCheckShowTempRange={jest.fn()}
            />,
        )

        wrapper.find({
            label: 'csv',
        }).prop('onClick')()

        setTimeout(() => {
            expect(downloadDocument).toHaveBeenCalledWith({
                headers: {
                    'content-type': 'text/csv',
                },
                data: 'SERIAL_NUMBER,TIMESTAMP,AMBIENT_TEMPERATURE,INTERNAL_TEMPERATURE\n031-10279,16.08.2018 06:50,24.7,21.5',
            }, 'sensor_data_031-10279')
            done()
        })
    })

    it('should generate csv data', () => {
        const wrapper = shallow(
            <DownloadOptions
                showCsvButton
                showTempRange
                sensorData={[[
                    new Date('August 16, 2018 06:50:00'),
                    24.7,
                    null,
                    null,
                    21.5,
                    null,
                    null,
                ]]}
                onCheckShowTempRange={jest.fn()}
            />,
        )

        expect(wrapper).toMatchSnapshot()
    })
})
