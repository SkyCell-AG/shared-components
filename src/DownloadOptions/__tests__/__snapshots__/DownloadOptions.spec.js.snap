// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shared-components/DownloadOptions should generate csv data 1`] = `
<Card
  className=""
  contentClass=""
  fullHeight={false}
  icon=""
  status=""
  title="Options"
  titleClass=""
>
  <div
    className="makeStyles-containerBtn-1"
  >
    <Button
      className="makeStyles-btnControl-2"
      disabled={false}
      icon={null}
      isUpperCase={true}
      label="csv"
      onClick={[Function]}
      saving={false}
      secondary={false}
    />
  </div>
</Card>
`;

exports[`shared-components/DownloadOptions should match snapshot 1`] = `
<Card
  className=""
  contentClass=""
  fullHeight={false}
  icon=""
  status=""
  title="Options"
  titleClass=""
>
  <div
    className="makeStyles-containerBtn-1"
  />
</Card>
`;

exports[`shared-components/DownloadOptions should match snapshot with all buttons 1`] = `
<Card
  className=""
  contentClass=""
  fullHeight={false}
  icon=""
  status=""
  title="Options"
  titleClass=""
>
  <TemperatureRange
    onCheckShowTempRange={[MockFunction]}
    showTempRange={true}
    title="Temperature Range"
  />
  <Radio
    className="makeStyles-selectedTemperatureRange-3"
    compact={true}
    disabled={false}
    name="selectedTemperatureRange"
    onChange={[Function]}
    options={Array []}
    showInColumns={false}
    small={false}
    title=""
    value=""
  />
  <div
    className="makeStyles-containerBtn-1"
  >
    <Button
      className="makeStyles-btnControl-2"
      disabled={false}
      icon={null}
      isUpperCase={true}
      label="pdf"
      onClick={[Function]}
      saving={false}
      secondary={false}
    />
    <Button
      className="makeStyles-btnControl-2"
      disabled={false}
      icon={null}
      isUpperCase={true}
      label="csv"
      onClick={[Function]}
      saving={false}
      secondary={false}
    />
  </div>
</Card>
`;
